### PR TITLE
Added DepthFirstSearch.java and DepthFirstTestSearch.java

### PR DESCRIPTION
**Pull request by abircb**
Implemented the DepthFirstSearch in Java by creating a Binary Tree class and adding appropriate 'add' and 'get' methods. Also added a JUnit test :)

### CODE REVIEW

File: src/main/java/com/search/DepthFirstSearch.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+ * What to do when adding or removing elements? Sort again? No.
+ * Create efficient data structures to maintain sorted sequences, and search in them
+ * Key example: binary sorted tree, allowing O(log N) insert, remove and lookup.
+   
+   In comes, depth-first search
+ * Worst-case performance	O(n)
+ * Best-case performance	O(1)
+ * Average performance	   	O(n)
+ * 
+ * @author abir (https://github.com/abircb)
```
Comment (yanglbme) [inline src/main/java/com/search/DepthFirstSearch.java]: Please do not add a signature inside the code. See [Contribution Guidelines](https://github.com/TheAlgorithms/Java/tree/Development#contribution-guidelines) for details.

Comment (yanglbme) [reply src/main/java/com/search/DepthFirstSearch.java]:
```suggestion

```

File: src/main/java/com/search/DepthFirstSearch.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        	return tree.get(key);
+	}
+	
+}
+
+/**
+ * The BinaryTree class defines the structure of a binary tree
+ * Also contains a static nested class called TreeNode
+ * @param <T>
+ * @author abir
```
Comment (yanglbme) [inline src/main/java/com/search/DepthFirstSearch.java]:
```suggestion

```

File: src/main/java/com/search/DepthFirstSearch.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ */
+class BinaryTree<T extends Comparable<T>> {
+	
+	private TreeNode<T> root;
+	
+	/**
+	 * @author abir
```
Comment (yanglbme) [inline src/main/java/com/search/DepthFirstSearch.java]:
```suggestion

```

File: src/main/java/com/search/DepthFirstSearch.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+				}
+			}
+		}
+		
+		/**
+		 * @param key
+		 * @return the tree node corresponding to the key
+		 */
+		private TreeNode<P> find(P key) {
+			if(key.compareTo(this.key) == 0) return this;
```
Comment (yanglbme) [inline src/main/java/com/search/DepthFirstSearch.java]: Add single space after the `if`
```suggestion
			if (key.compareTo(this.key) == 0) return this;
```

File: src/main/java/com/search/DepthFirstSearch.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+			else if(key.compareTo(this.key) < 0) {
```
Comment (yanglbme) [inline src/main/java/com/search/DepthFirstSearch.java]:
```suggestion
			else if (key.compareTo(this.key) < 0) {
```

File: src/main/java/com/search/DepthFirstSearch.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+				if(this.left == null) return null;
```
Comment (yanglbme) [inline src/main/java/com/search/DepthFirstSearch.java]:
```suggestion
				if (this.left == null) return null;
```

Same to the below.

File: src/test/java/com/search/DepthFirstSearchTest.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
+package src.test.java.com.search;
+
+import org.junit.Assert;
+import org.junit.Test;
+import src.main.java.com.search.DepthFirstSearch;
+import src.main.java.com.search.BinaryTree;
```
Comment (yanglbme) [inline src/test/java/com/search/DepthFirstSearchTest.java]: 'src.main.java.com.search.BinaryTree' is not public in 'src.main.java.com.search'. Cannot be accessed from outside package.

File: src/test/java/com/search/DepthFirstSearchTest.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        tree1.add(2,2);
+        tree1.add(3,3);
+        tree1.add(4,4);
+        Assert.assertEquals("Incorrect index", 3, DepthFirstSearch.find(3, tree1));
+        Assert.assertEquals("Incorrect index", 1, DepthFirstSearch.find(1, tree1));
+        Assert.assertEquals("Incorrect index", null, DepthFirstSearch.find(0, tree1));
+        Assert.assertEquals("Incorrect index", null, DepthFirstSearch.find(8, tree1));
+        Assert.assertEquals("Incorrect index", null, DepthFirstSearch.find(-2, tree1));
+
+        BinaryTree<String> tree2 = new BinaryTree<String>();
```
Comment (yanglbme) [inline src/test/java/com/search/DepthFirstSearchTest.java]:
```suggestion
        BinaryTree<String> tree2 = new BinaryTree<>();
```

File: src/test/java/com/search/DepthFirstSearchTest.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import src.main.java.com.search.DepthFirstSearch;
+import src.main.java.com.search.BinaryTree;
+
+public class DepthFirstSearchTest {
+    @Test
+    public void testDepthFirstSearch() {
+
+        BinaryTree<Integer> tree1 = new BinaryTree<Integer>();
```
Comment (yanglbme) [inline src/test/java/com/search/DepthFirstSearchTest.java]:
```suggestion
        BinaryTree<Integer> tree1 = new BinaryTree<>();
```

**Comment by yanglbme** [comment]
Hi @abircb, I've revert the PR `Added DepthFirstSearch.java and DepthFirstTestSearch.java #737`

You can open a new pull request after fixing the problems.

**Comment by abircb** [comment]
> Hi @abircb, I've revert the PR `Added DepthFirstSearch.java and DepthFirstTestSearch.java #737`
> 
> You can open a new pull request after fixing the problems.

**Comment by abircb** [comment]
Hi @yanglbme, I've just opened a new pull request- please check it out. And thanks for all the valuable feedback :)

**Pull request merged**